aws/ec2.go

```diff
@@ -0,0 +1,226 @@
+	if err != nil {
+		return "", err
+	}
+	log.Debugf("ec2-instance-id: %s", idDoc.InstanceID)
+	return idDoc.InstanceID, nil
+}
+
+func getEC2Region(ctx context.Context, svc ec2MetadaAPI) (string, error) {
+	region, err := svc.RegionWithContext(ctx)
+	if err != nil {
+		return "", err
+	}
+	log.Debugf("region: %s", region)
+	return region, nil
+}
+
+type ec2Client struct {
+	EC2Svc        ec2iface.EC2API
+	ec2InstanceId string
+}
+
+func newEC2Client(ctx context.Context) (*ec2Client, error) {
+	awsSession, err := awssession.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("error creating AWS session: %v", err)
+	}
+
+	metadataSvc := ec2metadata.New(awsSession)
+	if metadataSvc == nil {
+		return nil, fmt.Errorf("error connecting to EC2 Metadata service")
+	}
+
+	if !metadataSvc.AvailableWithContext(ctx) {
+		return nil, fmt.Errorf("EC2 metadata service is unavailable or not running on an EC2 instance")
+	}
+
+	region, err := getEC2Region(ctx, metadataSvc)
+	if err != nil {
+		return nil, fmt.Errorf("error getting ec2 region: %s", convertError(err))
+	}
+
+	instanceId, err := getEC2InstanceID(ctx, metadataSvc)
+	if err != nil {
+		return nil, fmt.Errorf("error getting ec2 instance-id: %s", convertError(err))
+	}
+
+	ec2Svc := ec2.New(awsSession, aws.NewConfig().WithRegion(region))
+	if ec2Svc == nil {
+		return nil, fmt.Errorf("error connecting to EC2 service")
+	}
+
+	return &ec2Client{
+		EC2Svc:        ec2Svc,
+		ec2InstanceId: instanceId,
+	}, nil
+}
+
+func (c *ec2Client) getEC2NetworkInterfaceId(ctx context.Context) (networkInstanceId string, err error) {
+	input := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{
+			aws.String(c.ec2InstanceId),
+		},
+	}
+
+	var out *ec2.DescribeInstancesOutput
+	for i := 0; i < retries; i++ {
+		out, err = c.EC2Svc.DescribeInstancesWithContext(ctx, input)
+		if err != nil {
+			if retriable(err) {
+				// if error is temporary, try again in a second.
+				time.Sleep(1 * time.Second)
+				log.WithField("instance-id", c.ec2InstanceId).Debug("retrying getting network-interface-id")
+				continue
+			}
+			return "", err
+		} else {
+			break
+		}
+	}
+
+	if out == nil || len(out.Reservations) == 0 {
+		return "", fmt.Errorf("no network-interface-id found for EC2 instance %s", c.ec2InstanceId)
+	}
+
+	var interfaceId string
+	for _, instance := range out.Reservations[0].Instances {
+		if len(instance.NetworkInterfaces) == 0 {
+			return "", fmt.Errorf("no network-interface-id found for EC2 instance %s", c.ec2InstanceId)
+		}
+		// We are only modifying network interface with device-id-0 to update
+		// instance source-destination-check.
+		// An instance can have multiple interfaces and the API response can be
+		// out-of-order interface list. We compare the device-id in the
+		// response to make sure the right device is updated.
+		for _, networkInterface := range instance.NetworkInterfaces {
+			if *(networkInterface.Attachment.DeviceIndex) == deviceIndexZero {
```

Comment:
I'd be inclined to put a nil check here; don't want felix to crash if the AWS API returns something unexpected.